x-pack/plugins/lens/public/indexpattern_plugin/indexpattern.tsx

```diff
@@ -40,80 +44,103 @@ export interface IndexPatternPersistedState {
   currentIndexPattern: string;
 
   columnOrder: string[];
-  columns: {
-    [columnId: string]: IndexPatternColumn;
-  };
+  columns: Record<string, IndexPatternColumn>;
 }
 
 export type IndexPatternPrivateState = IndexPatternPersistedState & {
-  indexPatterns: { [id: string]: IndexPattern };
+  indexPatterns: Record<string, IndexPattern>;
 };
 
-// Not stateful. State is persisted to the frame
-export const indexPatternDatasource: Datasource<
-  IndexPatternPrivateState,
-  IndexPatternPersistedState
-> = {
-  async initialize(state?: IndexPatternPersistedState) {
-    // TODO: Make fetch request to load indexPatterns from saved objects
-    if (state) {
+export function getIndexPatternDatasource(chrome: Chrome, toastNotifications: ToastNotifications) {
+  // Not stateful. State is persisted to the frame
+  const indexPatternDatasource: Datasource<IndexPatternPrivateState, IndexPatternPersistedState> = {
+    async initialize(state?: IndexPatternPersistedState) {
+      const indexPatternObjects = await getIndexPatterns(chrome, toastNotifications);
+      const indexPatterns: Record<string, IndexPattern> = {};
+
+      if (indexPatternObjects) {
+        indexPatternObjects.forEach((obj: Partial<IndexPattern>) => {
+          indexPatterns[obj.id!] = obj as IndexPattern;
+        });
+      }
+
+      // TODO: Make fetch request to load indexPatterns from saved objects
+      if (state) {
+        return {
+          ...state,
+          indexPatterns,
+        };
+      }
       return {
-        ...state,
-        indexPatterns: {},
+        currentIndexPattern: indexPatternObjects ? (indexPatternObjects[0] as IndexPattern).id : '',
```

Comment:
nit: Should we go with `null` here to make clear there is no current index pattern?

Reply:
I'm going to be changing this in the next PR anyway, which is adding some basic functionality around switching index patterns. So I'll wait on that.